applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,576 @@
+            return True
+        elif constraint_value >= self.constraint_buffer_variables[identifier]["lower_buffer_value"]:
+            return True
+        else:
+            return False
+    # --------------------------------------------------------------------------
+    def __updateBufferZone(self):
+    # adapt the buffer zones for zig-zagging, too much or too little correction
+        for constraint in self.constraints:
+
+            identifier = constraint["identifier"].GetString()
+            constraint_value = self.communicator.getStandardizedValue(identifier)
+
+            if self.optimization_iteration > 3:
+                delta_g_1 = constraint_value - self.constraint_buffer_variables[identifier]["constraint_value-1"]
+                delta_g_2 = self.constraint_buffer_variables[identifier]["constraint_value-1"] -self.constraint_buffer_variables[identifier]["constraint_value-2"]
+                delta_g_3 = self.constraint_buffer_variables[identifier]["constraint_value-2"] -self.constraint_buffer_variables[identifier]["constraint_value-3"]
+
+                if delta_g_1*delta_g_2 < 0 and delta_g_2*delta_g_3 < 0:
+                    self.constraint_buffer_variables[identifier]["buffer_size_factor"] += abs(self.constraint_buffer_variables[identifier]["buffer_value"]-self.constraint_buffer_variables[identifier]["buffer_value-1"])
+
+            if self.optimization_iteration > 1:
+                delta_g = constraint_value - self.constraint_buffer_variables[identifier]["constraint_value-1"]
+                print("constraint_value = ", constraint_value)
+                print("delta_g = ", delta_g)
+                print("constraint_value - 1 = ", self.constraint_buffer_variables[identifier]["constraint_value-1"])
+                if delta_g >= 0.0 and self.constraint_buffer_variables[identifier]["constraint_value-1"] > 0:
+                    self.constraint_buffer_variables[identifier]["central_buffer_value"] -= self.constraint_buffer_variables[identifier]["constraint_value-1"]
+                elif delta_g <= 0.0 and self.constraint_buffer_variables[identifier]["constraint_value-1"] < 0:
+                    self.constraint_buffer_variables[identifier]["central_buffer_value"] += self.constraint_buffer_variables[identifier]["constraint_value-1"]
+                    self.constraint_buffer_variables[identifier]["central_buffer_value"] = max(self.constraint_buffer_variables[identifier]["central_buffer_value"],0.0)
+
+            self.constraint_buffer_variables[identifier]["constraint_value-3"] = self.constraint_buffer_variables[identifier]["constraint_value-2"]
+            self.constraint_buffer_variables[identifier]["constraint_value-2"] = self.constraint_buffer_variables[identifier]["constraint_value-1"]
+            self.constraint_buffer_variables[identifier]["constraint_value-1"] = constraint_value
+    # --------------------------------------------------------------------------
+
+    def __normOutputGradients(self):
```

Comment:
can you find a better name for this function?

does it norm**alize** gradients? what does "output" mean here?

Reply:
Why do you need to normalize the gradients after computation of shape update?